crates/deltalake/examples/basic_operations_kms_encryption.rs

```diff
@@ -0,0 +1,671 @@
+            ])),
+        ],
+    )
+    .unwrap();
+    ctx.read_batch(batch).unwrap()
+}
+
+async fn merge_table(uri: &str, table_encryption: &TableEncryption) -> Result<(), DeltaTableError> {
+    let ops = ops_with_crypto(uri, table_encryption).await?;
+
+    let schema = get_table_schema();
+    let source = merge_source(schema);
+
+    let (table, _metrics) = ops
+        .merge(source, col("target.int").eq(col("source.int")))
+        .with_source_alias("source")
+        .with_target_alias("target")
+        .when_not_matched_by_source_delete(|delete| delete)
+        .unwrap()
+        .await
+        .unwrap();
+
+    let expected = vec![
+        "+-----+--------+----------------------------+",
+        "| int | string | timestamp                  |",
+        "+-----+--------+----------------------------+",
+        "| 10  | A      | 1970-01-01T00:08:20.012305 |",
+        "| 10  | A      | 1970-01-01T00:08:20.012305 |",
+        "+-----+--------+----------------------------+",
+    ];
+
+    let (_table, stream) = DeltaOps(table).load().await?;
+    let data: Vec<RecordBatch> = collect_sendable_stream(stream).await?;
+
+    // println!("{data:?}");
+
+    assert_batches_sorted_eq!(&expected, &data);
+    Ok(())
+}
+
+async fn optimize_table_z_order(
+    uri: &str,
+    table_encryption: &TableEncryption,
+) -> Result<(), DeltaTableError> {
+    let ops = ops_with_crypto(uri, table_encryption).await?;
+    let (_table, metrics) = ops
+        .optimize()
+        .with_type(OptimizeType::ZOrder(vec![
+            "timestamp".to_string(),
+            "int".to_string(),
+        ]))
+        .await?;
+    println!("\nOptimize Z-Order:\n{metrics:?}\n");
+    Ok(())
+}
+
+async fn optimize_table_compact(
+    uri: &str,
+    table_encryption: &TableEncryption,
+) -> Result<(), DeltaTableError> {
+    let ops = ops_with_crypto(uri, table_encryption).await?;
+    let (_table, metrics) = ops.optimize().with_type(OptimizeType::Compact).await?;
+    println!("\nOptimize Compact:\n{metrics:?}\n");
+    Ok(())
+}
+
+/*
+I guess this test isn't needed since checkpoints only summarize what files to use.
+ */
+async fn checkpoint_table(
+    uri: &str,
+    table_encryption: &TableEncryption,
+) -> Result<(), DeltaTableError> {
+    let table_location = uri;
+    let table_path = PathBuf::from(table_location);
+    let log_path = table_path.join("_delta_log");
+
+    let ops = ops_with_crypto(uri, table_encryption).await?;
+    let table: DeltaTable = ops.into();
+    let version = table.version();
+
+    // Write a checkpoint
+    checkpoints::create_checkpoint(&table, None).await.unwrap();
+
+    // checkpoint should exist
+    let filename = format!(
+        "00000000000000000{:03}.checkpoint.parquet",
+        version.unwrap()
+    );
+    let checkpoint_path = log_path.join(filename);
+    assert!(checkpoint_path.as_path().exists());
+
+    Ok(())
+}
+
+async fn round_trip_test() -> Result<(), deltalake::errors::DeltaTableError> {
```

Comment:
As well as having an example, it would be good to have unit tests for each type of operation that check that the table files are all encrypted after the operation is complete.